### DISCUSSION TITLE
Working with .Net 6.0 with other services

### DISCUSSION THREAD

**Discussion by Artoryn** [mariuszkerl/AspNetCoreSubdomain#70] [General]
Just wanted to make the note here as I had been working with this for a while and finally figured out what was wrong. I also possibly missed something, but, make sure that the `AddSubdomains()` and `AddMvc(x => x.EnableEndpointRouting = false)` are added as the first items to the builder before any other framework services. I had this code (`AddSubdomains()` and `AddMvc(...)`) lower towards the end of the builder section and it would not properly generate the action links. The subdomain routing was working when entering the subdomain into the browser's address bar, but not with the action link generation.

For .Net 6.0, if you are using the default Program.cs file, it probably should look something like this:
```
using ...

var builder = WebApplication.CreateBuilder(args);

// Do the calls for this Package first
builder.Services.AddSubdomain();
builder.Services.AddMvc(x => x.EnableEndpointRouting = false);

// Add/Configure your services
builder.Services.Configure...

var app = builder.Build();

// app.Use...

var hostnames = new[] { "localhost:54575" };
    
app.UseMvc(routes =>
{
    routes.MapRoute(
        hostnames,
        "NormalRoute",
        "{action}",
        new { controller = "SomeController" });

    routes.MapSubdomainRoute(
        hostnames,
        "SubdomainRoute",
        "{controller}", //that's subdomain parameter, it can be anything
        "{action}",
        new { controller = "Home", action = "Action1" });
)};

app.Run();
```

I hope this helps someone.

**Comment by mariuszkerl** [comment]
Hi @ceolson96 .
Thanks for your input. I will verify and improve description on the main page.

**Reply by mariuszkerl** [reply]
Hmm unluckily working as expected on mine. (net core3, net 6). I've added healthchecks and dependency injection on top of the configuration. Links are generated correctly.

**Reply by Artoryn** [reply]
To try and help, I have generated an example project that on my system does replicate the issue.

When the mouse is on the link displayed 'Test Link' which should show 'test.localhost:<Port>' but doesn't.
![image](https://user-images.githubusercontent.com/42783832/155244479-bff519c5-6852-44c5-a0f9-d037aa05d1c0.png)

But when the builder is at the top of the method, the link displayed 'Test Link' does show 'test.localhost:<port>'.
![image](https://user-images.githubusercontent.com/42783832/155244649-9fd424ae-33af-47be-b94b-c9c573fa2d90.png)

Do note that the project is 'Asp.Net Core Web App (MVC)' with individual accounts.
To view this test application, please see the below link.
[https://github.com/ceolson96/ANCSDExampleProject](https://github.com/ceolson96/ANCSDExampleProject)

I hope this helps.

**Reply by mariuszkerl** [reply]
Hi @ceolson96 ,

thanks for the sample, that's really helpful. Two things to the code:
1) You don't need `AddControllerWithViews` method if you use `AddMvc`. I've found [this link](https://dotnettutorials.net/lesson/difference-between-addmvc-and-addmvccore-method/) really helpful.

2) `AddDefaultIdentity` is adding some built in routing and is the source of the problem. We can partially fix it by moving `AddSubdomains` method to the top, I say _partially_ since identity routing will be working on subdomain path which to me would be unexpected. I will add known bug for it. Also if you want to solve your issue I think you were right with moving to the top approach (you only have to move method mentioned before.

**Reply by Artoryn** [reply]
With the Identity, I did get it working with a subdomain, however, I converted the Identity from RazorPages to MVC as well as changed the `AddDefaultIdentity` to be `AddIdentity`. Kind of a workaround on Identity as it is time-consuming to change it over to MVC from my experience, but can work.
![image](https://user-images.githubusercontent.com/42783832/155440321-0735fd71-8cd0-44a1-9c3e-10668bfae5de.png)
